pennylane/transforms/op_transforms.py

```diff
@@ -0,0 +1,481 @@
+# limitations under the License.
+"""
+This module contains the @op_transform decorator.
+"""
+# pylint: disable=protected-access
+import functools
+import inspect
+import os
+import warnings
+
+import pennylane as qml
+
+
+class OperationTransformError(Exception):
+    """Raised when there is an error with op_transform logic"""
+
+
+def _make_tape(obj, wire_order, *args, **kwargs):
+    """Given an input object, which may be:
+
+    - an object such as a tape or a operation, or
+    - a callable such as a QNode or a quantum function
+      (alongside the callable arguments ``args`` and ``kwargs``),
+
+    this function constructs and returns the tape/operation
+    represented by the object.
+
+    The ``wire_order`` argument determines whether a custom wire ordering
+    should be used. It not provided, the wire ordering defaults to the
+    objects wire ordering accessed via ``obj.wires``.
+    """
+    if isinstance(obj, qml.QNode):
+        # user passed a QNode, get the tape
+        obj.construct(args, kwargs)
+        tape = obj.qtape
+        wires = obj.device.wires
+
+    elif isinstance(obj, qml.tape.QuantumTape):
+        # user passed a tape
+        tape = obj
+        wires = tape.wires
+
+    elif inspect.isclass(obj) and issubclass(obj, qml.operation.Operator):
+        with qml.tape.stop_recording():
+            tape = obj(*args, **kwargs)
+
+        wires = tape.wires
+
+    elif callable(obj):
+        # user passed something that is callable but not a tape or QNode.
+        tape = qml.transforms.make_tape(obj)(*args, **kwargs)
+        wires = tape.wires
+
+        # raise exception if it is not a quantum function
+        if len(tape.operations) == 0 and len(tape.measurements) == 0:
+            raise OperationTransformError("Quantum function contains no quantum operations")
+
+    # if no wire ordering is specified, take wire list from tape/device
+    wire_order = wires if wire_order is None else qml.wires.Wires(wire_order)
+
+    # check that all wire labels in the circuit are contained in wire_order
+    if not set(tape.wires).issubset(wire_order):
+        raise OperationTransformError(
+            f"Wires in circuit {tape.wires.tolist()} are inconsistent with "
+            f"those in wire_order {wire_order.tolist()}"
+        )
+
+    return tape, wire_order
+
+
+class op_transform:
+    r"""Class for registering an operator transform that takes one or more operators,
+    and returns a classical representation.
+
+    Using ``op_transform`` is not necessary in most cases; simply define a
+    standard Python function that accepts an operator and returns the
+    computed quantity.
+
+    However, this registration class is useful if you wish to easily create
+    a function that:
+
+    - Supports datastructures that may contain multiple operations, such as
+      a tape, QNode, or qfunc.
+
+    - Supports being used with a functional transform UI.

```

Comment:
I really like the explanation in the PR description, and find it more useful that this one. I wonder if there is an easier way to tell a developer if she needs this transform, and what it does? Maybe with a few in-line examples already up here?